packages/runtime/container-runtime/src/opDecompressor.ts

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { decompress } from "lz4js";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { assert, IsoBuffer } from "@fluidframework/common-utils";
+import { CompressionAlgorithms } from ".";
+
+/**
+ * State machine that "unrolls" contents of compressed batches of ops after decompressing them.
+ * This class relies on some implicit contracts defined below:
+ * 1. A compressed batch will have its first message with batch metadata set to true and compressed set to true
+ * 2. Messages in the middle of a compressed batch will have neither batch metadata nor the compression property set
+ * 3. The final message of a batch will have batch metadata set to false
+ * 4. An individually compressed op will have undefined batch metadata and compression set to true
+ */
+export class OpDecompressor {
+    private activeBatch = false;
+    private rootMessage: ISequencedDocumentMessage | undefined;
+    private processedCount = 0;
+
+    constructor() {
+
+    }
+
+    public processMessage(message: ISequencedDocumentMessage): ISequencedDocumentMessage {
+        // Beginning of a compressed batch
+        if (message.metadata?.batch === true
+            && (message.metadata?.compressed || (message as any).compression === CompressionAlgorithms.lz4)) {
```

Comment:
It feels like you want to assert, not runtime check that it's lz4. I.e. if it's something else, we are screwed - there is no code to parse it.

Reply:
Please consider pushing small change into last release branch that will inject immediate failure if client observes compressed ops. Or at least validating that when old client gets compressed ops, it fails (already) in some reasonable way that we can identify from logging (though better user message would be nice).
Overall, we should think about adding unique errorType for cases where client detects it does not know format of the op (i.e. convert some asserts into DataProcessingError cases with unique payload that tells us and customers what actually happened).

Reply:
I've changed this to a runtime check to see if the top level property is defined and then asserting that it's lz4 so that the metadata only path works (if clients want to enable it before ordering service is ready).